gridded_etl_tools/utils/extractor.py

```diff
@@ -0,0 +1,363 @@
+                )
+                self.dm.info(f"Finished downloading {informative_id}")
+                return True
+            except Exception as e:
+                # Increase delay time after each failure
+                retry_delay = counter * 30
+                self.dm.info(
+                    f"Encountered exception {e} for {informative_id}, retrying after {retry_delay} seconds"
+                    f" , attempt {counter}"
+                )
+                counter += 1
+        else:
+            self.dm.info(f"Couldn't find or download a remote file for {informative_id}")
+            raise FileNotFoundError(f"Too many ({counter}) failed download attempts from server")
+
+
+class FTPExtractor(Extractor):
+    """
+    Create an object that provides an interface to a climate data source's FTP server. The object is able to open its
+    connection within a context manager, navigate to specific working directory, match files located in subdirectories,
+    and fetch files to a given destination folder.
+
+    Currently only anonymous FTP access is supported.
+    """
+
+    ftp: ftplib.FTP
+    host: str
+
+    # Used to limit retrieval to a single thread at a time
+    semaphore: multiprocessing.synchronize.Semaphore = multiprocessing.Semaphore()
+
+    def __enter__(self) -> FTPExtractor:
+        """
+        Open a connection to the FTP server at this object's given source from within a context manager. When creating
+        a context, pass the host address unless `FTPExtractor.host` was already set.
+
+        Example
+        -------
+        my_extractor = FTPExtractor(my_dataset_manager)
+        with my_extractor("ftp.cdc.noaa.gov") as extractor:
+            # get source files
+
+        Returns
+        -------
+        FTPConnection
+            this object
+
+        Raises
+        ------
+        ValueError
+            If FTPExtractor.
+        """
+        if not hasattr(self, "host"):
+            raise ValueError("FTPExtractor must have a host parameter to open connection")
+        else:
+            self.dm.info(f"Opening a connection to {self.host}")
+            self.ftp = ftplib.FTP(self.host)
+            self.ftp.login()
+            return self
+
+    def __exit__(self, *exception):
+        """
+        Close the connection with the FTP server at this object's given source. This will be called automatically when
+        exiting the connection context manager.
+
+        Parameters
+        ----------
+        *exception
+            Exception information passed automatically by Python
+        """
+        self.ftp.close()
+        self.dm.info(f"Closed connection to {self.host}")
+
+    def __call__(self, host: str):
+        """
+        When called like a function, set the host parameter. See `FTPExtractor.__enter__`.
+
+        Parameters
+        ----------
+        host
+            Address to connect to for source data
+        """
+        self.host = host
+        return self
```

Comment:
I think the more readable way to do this is to have `host` be a parameter to `__enter__` and set the instance variable there

Reply:
Are you sure? That's what I tried originally, and I couldn't pass an argument. I ended up using the method [here](https://stackoverflow.com/a/68300883) which mentions `__enter__` doesn't accept arguments.

Reply:
My mistake - I was mixing up args to `__enter__` with the technique where you pass the args in `__init__` in a one liner with the `with` statement.